README.md
SellJunk sells all junk to the vendors. This includes all grey items and user specified items, that can be added to list of exceptions. It can either sell these items automatically at each vendor or you can do it manually by clicking "Sell Junk" button on each vendor's frame. You can also see how much gold you made from selling the trash. It can also destroy all junk instead of selling it, if you're far from vendors.

### Commands:

* **/sj** 
  * Will open config window
* **/sj add [item link]** 
  * Will add given item to exception list (to create itemlink Shift-Click any item in inventory)
* **/sj destroy**
  * Will destroy all junk in inventory
* **/sj destroy [x]**
  * Will destroy x pieces of junk in inventory

### Localization:
If you'd like to add/update localizations, do so [here](http://www.wowace.com/projects/sell-junk/localization/).

### Bugs / Suggestions:
If you have encountered any problem, or if you think there's something missing. Feel free to [open an issue](https://github.com/Beast-Masters-addons/SellJunk/issues).

### FAQ
* **What is an exception list?**
If you want the addon to sell other than grey quality items, or you want to not sell some of those grey items, then exception list is the thing you are looking for.
* **How do I add item to exception list?**
Type "/selljunk" and then type either name of the item, or just part of the name (but be aware it might sell items with similar names), or just drag the item from inventory into the input field.
* **What if I want to stop it from selling some grey quality item?**
Just add this item to exception list, and it will always skip it.
* **Can it destroy soulbound items, that you can't use (e.g. plate items on mage)?**
Nope, and it never will support it. The aim is to stay as lightweight as possible.

The addon was originally released on [CurseForge](https://www.curseforge.com/wow/addons/sell-junk) by [cbenneh](https://www.curseforge.com/members/cbenneh/) with GPL license, which allowed a long-time user to continue the development after being abandoned by the original author.

changelog.txt
3.0.19:
  - re-release due to packaging problem

3.0.18:
  - update: updated for 8.1

3.0.17:
  - update: fixes for 8.0

3.0.16:
  - update: updated for 8.0

3.0.15:
  - update: updated for patch 7.2

3.0.14:
  - update: updated for patch 7.1

3.0.13:
  - update: tweak output of gold

3.0.12:
  - update: updated for patch 7.0.3

3.0.11:
  - update: updated for patch 6.0.2

3.0.10:
  - update: updated for patch 5.4

3.0.9:
  - update: updated for patch 5.3

3.0.8:
  - update: when GnomishVendorShrinker is loaded, put SellJunk button to top right corner (hopefully nothing can ever be there!)

3.0.7:
  - update: when GnomishVendorShrinker is loaded, put SellJunk button to bottom left corner

3.0.6:
  - update: added Italian translation

3.0.5:
  - update: updated for patch 5.0.4

3.0.4:
  - update: added new translations

3.0.3:
  - update: updated for patch 4.3

3.0.2:
  - update: updated for patch 4.2

3.0.1:
  - update: updated for patch 4.1

3.0:
  - Rewritten to improve performance
  - feature: merged global and character exceptions into single list
  - update: updated for patch 4.0.1

2.6.3:
  - feature: added commands '/sj add <itemlink>' and '/sj rem <itemlink>', that will add/remove given item to global exceptions
  - bugfix: it should now skip items, that cannot be sell to a vendor as there are quite a few grey ones

2.6.2:
  - bugfix: '/sj destroy <number>' did not work properly

2.6.1:
  - bugfix: ticket #13

2.6:
  - feature: added command '/sj destroy', that will delete all junk from bags and '/sj destroy x', that will delete up to x pieces of junk

2.5.1:
  - update: updated for patch 3.3
  - update: spanish (esES) locale

2.5:
  - feature: option to hide the 'item sold' spam
  - update: french (frFR) locale

2.4.3:
  - update: korean (koKR) locale

2.4.2:
  - update: more efficient code for getting item price

2.4.1:
  - update: updated for patch 3.2

2.4:
  - feature: option to clear global or character specific exceptions
  - removed: option to list all exceptions, it was useless anyway

2.3.2:
  - feature: french (frFR) locale
  - update: various locales

2.3.1:
	- feature: german (deDE) locale

2.3:
	- feature: option to show gold gain from selling trash
	- bugfix: was possible to have same item many times on exception list


SellJunk.lua
SellJunk = LibStub("AceAddon-3.0"):NewAddon("SellJunk", "AceConsole-3.0","AceEvent-3.0")
local addon	= LibStub("AceAddon-3.0"):GetAddon("SellJunk")
local AceConfigRegistry = LibStub("AceConfigRegistry-3.0")
local AceConfigDialog = LibStub("AceConfigDialog-3.0")

local addonName, addonTable = ...
local L = addonTable.L
local _

addon.optionsFrame = {}
local options = nil

addon.sellButton = CreateFrame("Button", nil, MerchantFrame, "OptionsButtonTemplate")

if IsAddOnLoaded("GnomishVendorShrinker") then
  addon.sellButton:SetPoint("TOPRIGHT", -23, 0)
else
  addon.sellButton:SetPoint("TOPLEFT", 60, -32)
end

addon.sellButton:SetText(L["Sell Junk"])
addon.sellButton:SetScript("OnClick", function() SellJunk:Sell() end)

-- upvalues
local floor = floor
local mod = mod
local string_find = string.find
local pairs = pairs
local wipe = wipe
local DeleteCursorItem = DeleteCursorItem
local GetContainerItemInfo = GetContainerItemInfo
local GetItemInfo = GetItemInfo
local PickupContainerItem = PickupContainerItem
local PickupMerchantItem = PickupMerchantItem


function addon:OnInitialize()
  self:RegisterChatCommand("selljunk", "HandleSlashCommands")
  self:RegisterChatCommand("sj", "HandleSlashCommands")

  self.db = LibStub("AceDB-3.0"):New("SellJunkDB")
  self.db:RegisterDefaults({
    char = {
      auto = false,
			max12 = true,
			printGold = true,
      showSpam = true
    },
    global = {
      exceptions = {},
    }
  })

  self:PopulateOptions()
  AceConfigRegistry:RegisterOptionsTable("SellJunk", options)
  addon.optionsFrame = AceConfigDialog:AddToBlizOptions("SellJunk", nil, nil, "general")
end

function addon:OnEnable()
  self:RegisterEvent("MERCHANT_SHOW")
	self.total = 0
end

function addon:MERCHANT_SHOW()
  if addon.db.char.auto then
    self:Sell()
  end
end

function addon:AddProfit(profit)
	if profit then
		self.total = self.total + profit
	end
end

-------------------------------------------------------------
-- Sells items:                                            --
--   - grey quality, unless it's in exception list         --
--   - better than grey quality, if it's in exception list --
-------------------------------------------------------------
function addon:Sell()
	local limit = 0
  local currPrice
  local showSpam = addon.db.char.showSpam
  local max12 = addon.db.char.max12

  for bag = 0,4 do
    for slot = 1,GetContainerNumSlots(bag) do
      local item = GetContainerItemLink(bag,slot)
      if item then
				-- is it grey quality item?
        local grey = string_find(item,"|cff9d9d9d")

        if (grey and (not addon:isException(item))) or ((not grey) and (addon:isException(item))) then
          currPrice = (select(11, GetItemInfo(item)) or 0) * select(2, GetContainerItemInfo(bag, slot))
          -- this should get rid of problems with grey items, that cant be sell to a vendor
          if currPrice > 0 then
            addon:AddProfit(currPrice)
            PickupContainerItem(bag, slot)
            PickupMerchantItem()
            if showSpam then
              self:Print(L["Sold"]..": "..item)
            end

            if max12 then
              limit = limit + 1
              if limit == 12 then
                return
              end
            end
          end
        end
      end
    end
  end

	if self.db.char.printGold then
		self:PrintGold()
	end
	self.total = 0
end

-------------------------------------------------------------
-- Destroys items:                                         --
--   - grey quality, unless it's in exception list         --
--   - better than grey quality, if it's in exception list --
-------------------------------------------------------------
function addon:Destroy(count)
  local limit = 9001 -- it's over NINE THOUSAND!!!
  if count ~= nil then
    limit = count
  end

  local showSpam = addon.db.char.showSpam

  for bag = 0,4 do
    for slot = 1,GetContainerNumSlots(bag) do
      local item = GetContainerItemLink(bag,slot)
      if item then
				-- is it grey quality item?
        local grey = string_find(item,"|cff9d9d9d")

        if (grey and (not addon:isException(item))) or ((not grey) and (addon:isException(item))) then
          PickupContainerItem(bag, slot)
					DeleteCursorItem()
          if showSpam then
            self:Print(L["Destroyed"]..": "..item)
          end
          limit = limit - 1
          if limit == 0 then
            break
          end
        end
      end
    end
    if limit == 0 then
      break
    end
  end

	if self.db.char.printGold then
		self:PrintGold()
	end
	self.total = 0
end

function addon:PrintGold()
  if self.total > 0 then
		self:Print(L["Gained"]..": "..GetMoneyString(self.total))
	end
end

function addon:Add(link)

	-- remove all trailing whitespace
	link = strtrim(link)

	-- extract name from an itemlink
  local found, _, name = string_find(link, "^|c%x+|H.+|h.(.*)\].+")

	-- if it's not an itemlink, guess it's name of an item
	if not found then
		name = link
	end

  local exceptions = self.db.global.exceptions
  for k,v in pairs(exceptions) do
    if v == name or v == link then
      return
    end
  end

  -- append name of the item to global exception list
  exceptions[#exceptions + 1] = name
  self:Print(L["Added"] .. ": " .. link)
end

function addon:Rem(link)
	-- remove all trailing whitespace
	link = strtrim(link)

	-- extract name from an itemlink
  local isLink, _, name = string_find(link, "^|c%x+|H.+|h.(.*)\].+")

	-- if it's not an itemlink, guess it's name of an item
	if not isLink then
		name = link
	end

  -- looping through exceptions
  local found = false
  local exception
  local exceptions = self.db.global.exceptions
  for k,v in pairs(exceptions) do
    found = false
    -- comparing exception list entry with given name
    if v:lower() == name:lower() then
      found = true
    end

    -- extract name from itemlink (only for compatibility with old saved variables)
    isLink, _, exception = string_find(v, "^|c%x+|H.+|h.(.*)\].+")
    if isLink then
      -- comparing exception list entry with given name
      if exception:lower() == name:lower() then
        found = true
      end
    end

    if found then
      if exceptions[k + 1] then
        exceptions[k] = exceptions[k + 1]
      else
        exceptions[k] = nil
      end
      self:Print(L["Removed"]..": "..link)
      break
    end
  end
end

function addon:isException(link)
	local exception = nil

	-- extracting name of an item from the itemlink
	local isLink, _, name = string_find(link, "^|c%x+|H.+|h.(.*)\].+")

	-- it's not an itemlink, so guess it's name of the item
	if not isLink then
		name = link
	end

  local exceptions = self.db.global.exceptions
	if exceptions then

		-- looping through global exceptions
		for k,v in pairs(exceptions) do

			-- comparing exception list entry with given name
			if v:lower() == name:lower() then
				return true
			end

			-- extract name from itemlink (only for compatibility with old saved variables)
			isLink, _, exception = string_find(v, "^|c%x+|H.+|h.(.*)\].+")
			if isLink then
				-- comparing exception list entry with given name
				if exception:lower() == name:lower() then
					return true
				end
			end
		end
	end

	-- item not found in exception list
	return false
end

function addon:ClearDB()
  wipe(self.db.global.exceptions)
  self:Print(L["Exceptions succesfully cleared."])
end

function addon:HandleSlashCommands(input)
  local arg1, arg2 = self:GetArgs(input, 2, 1, input)
  if arg1 == 'destroy' then
    self:Destroy(arg2)
  elseif arg1 == 'add' and arg2 ~= nil then
    if arg2:find('|Hitem') == nil then
      self:Print(L["Command accepts only itemlinks."])
    else
      self:Add(arg2, true)
    end
  elseif (arg1 == 'rem' or arg1 == 'remove') and arg2 ~= nil then
    if arg2:find('|Hitem') == nil then
      self:Print(L["Command accepts only itemlinks."])
    else
      self:Rem(arg2, true)
    end
  else
    InterfaceOptionsFrame_OpenToCategory(addon.optionsFrame)
  end
end

function addon:PopulateOptions()
	if not options then
		options = {
			order = 1,
			type  = "group",
			name  = "SellJunk",
			args  = {
				general = {
					order	= 1,
					type	= "group",
					name	= "global",
					args	= {
						divider1 = {
							order	= 1,
							type	= "description",
							name	= "",
						},
						auto = {
							order	= 2,
							type 	= "toggle",
							name 	= L["Automatically sell junk"],
							desc 	= L["Toggles the automatic selling of junk when the merchant window is opened."],
							get 	= function() return addon.db.char.auto end,
							set 	= function() self.db.char.auto = not self.db.char.auto end,
						},
						divider2 = {
							order	= 3,
							type	= "description",
							name	= "",
						},

						max12 = {
							order = 4,
							type  = "toggle",
							name  = L["Sell max. 12 items"],
							desc  = L["This is failsafe mode. Will sell only 12 items in one pass. In case of an error, all items can be bought back from vendor."],
							get 	= function() return addon.db.char.max12 end,
							set 	= function() self.db.char.max12 = not self.db.char.max12 end,
						},
						divider3 = {
							order	= 5,
							type	= "description",
							name	= "",
						},
						printGold = {
							order = 6,
							type  = "toggle",
							name  = L["Show gold gained"],
							desc  = L["Shows gold gained from selling trash."],
							get 	= function() return addon.db.char.printGold end,
							set 	= function() self.db.char.printGold = not self.db.char.printGold end,
						},
            divider4 = {
							order	= 7,
							type	= "description",
							name	= "",
						},
            showSpam = {
              order = 8,
              type  = "toggle",
              name  = L["Show 'item sold' spam"],
              desc  = L["Prints itemlinks to chat, when automatically selling items."],
              get   = function() return addon.db.char.showSpam end,
              set   = function() addon.db.char.showSpam = not addon.db.char.showSpam end,
            },
						divider5 = {
							order	= 9,
							type	= "header",
							name	= L["Clear exceptions"],
						},
						clearglobal = {
							order	= 10,
							type 	= "execute",
							name 	= L["Clear"],
              desc  = L["Removes all exceptions."],
							func 	= function() addon:ClearDB() end,
						},
						divider6 = {
							order	= 12,
							type	= "description",
							name	= "",
						},
						header1 = {
							order	= 13,
							type	= "header",
							name	= L["Exceptions"],
						},
						note1 = {
							order = 14,
							type 	= "description",
							name	= L["Drag item into this window to add/remove it from exception list"],
						},
						add = {
							order	= 15,
							type 	= "input",
							name 	= L["Add item"]..':',
							usage = L["<Item Link>"],
							get 	= false,
							set 	= function(info, v) addon:Add(v) end,
						},
						rem = {
							order	= 16,
							type 	= "input",
							name 	= L["Remove item"]..':',
							usage 	= L["<Item Link>"],
							get 	= false,
							set 	= function(info, v) addon:Rem(v) end,
						},
					}
				}
			}
		}
	end
end
